In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv('House Data.csv')

In [3]:
data.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [4]:
data.tail()

area_type   availability               location       size  \
13315        Built-up  Area  Ready To Move             Whitefield  5 Bedroom   
13316  Super built-up  Area  Ready To Move          Richards Town      4 BHK   
13317        Built-up  Area  Ready To Move  Raja Rajeshwari Nagar      2 BHK   
13318  Super built-up  Area         18-Jun        Padmanabhanagar      4 BHK   
13319  Super built-up  Area  Ready To Move           Doddathoguru      1 BHK   

       society total_sqft  bath  balcony  price  
13315  ArsiaEx       3453   4.0      0.0  231.0  
13316      NaN       3600   5.0      NaN  400.0  
13317  Mahla T       1141   2.0      1.0   60.0  
13318  SollyCl       4689   4.0      1.0  488.0  
13319      NaN        550   1.0      1.0   17.0

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


In [6]:
data.describe()

bath       balcony         price
count  13247.000000  12711.000000  13320.000000
mean       2.692610      1.584376    112.565627
std        1.341458      0.817263    148.971674
min        1.000000      0.000000      8.000000
25%        2.000000      1.000000     50.000000
50%        2.000000      2.000000     72.000000
75%        3.000000      2.000000    120.000000
max       40.000000      3.000000   3600.000000

In [7]:
    data.shape

(13320, 9)

Data have *13320 number of rows and 9 columns*

In [8]:
for i in data.columns:
    print(data[i].value_counts()) # we cal. the number of times the values occuring in a certain column.
    print("*"*20)

Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2025
Carpet  Area              87
Name: area_type, dtype: int64
********************
Ready To Move    10581
18-Dec             307
18-May             295
18-Apr             271
18-Aug             200
                 ...  
17-Jan               1
15-Jun               1
16-Nov               1
15-Dec               1
16-Jan               1
Name: availability, Length: 81, dtype: int64
********************
Whitefield              540
Sarjapur  Road          399
Electronic City         302
Kanakpura Road          273
Thanisandra             234
                       ... 
Chowdeshwari Layout       1
6th Block Jayanagar       1
Ashraya Layout            1
Mudalpalaya               1
B Y Raveshwara Nagar      1
Name: location, Length: 1305, dtype: int64
********************
2 BHK         5199
3 BHK         4310
4 Bedroom      826
4 BHK          591
3 Bedroom      547
1 BHK          538
2 Bedroom      329
5 Bedroom

Here we check that how many times does a single value occurs in the entire data.
Like, for the Super built-up Area : 8790, it means that this has occures 8790 times in its respective column, and so the others

In [9]:
print(data.isna().sum())
print("*"*20)
print(data.isnull().sum())

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64
********************
area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64


Here we can see that `society, balcony and bath` have NULL values. So we will drop them, coz `society` has so many NULL values, we do not need `balcony` column also, also we don't need `availiblity` column.

In [10]:
data.drop(columns = ['area_type', 'balcony', 'availability', 'society'], inplace=True)

In [11]:
data.describe()

# here we can see that we only have 2 numerical columns i.e. bath and price

bath         price
count  13247.000000  13320.000000
mean       2.692610    112.565627
std        1.341458    148.971674
min        1.000000      8.000000
25%        2.000000     50.000000
50%        2.000000     72.000000
75%        3.000000    120.000000
max       40.000000   3600.000000

## bath (bathroom) :
Here we can see that the **min - bathroom** have the value `1.00000`, this means that the bath have its min value as 1, and max value as 40.
The first quartile is of 2 bath, median is of 3, third quartile is of 3.

## Price:
Here we can see that the,
~ **max- price** of the bathroom os, 3600 lakhs
~ **min - price** = 8 lakhs
~  **mean - price** = 112.5 lakhs


In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13319 non-null  object 
 1   size        13304 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13247 non-null  float64
 4   price       13320 non-null  float64
dtypes: float64(2), object(3)
memory usage: 520.4+ KB



We can see that there are 13320 entries, whereas in the `location` we see that there are 13319 values, i.e. 1 value is missing.
Smilarly we can see that there are 13304, values in `size` column, that means that there are about 16 missing values.

# Filling up the Missing Values

In [13]:
# We'll check the count of the locations that has occured.
data['location'].value_counts()

Whitefield              540
Sarjapur  Road          399
Electronic City         302
Kanakpura Road          273
Thanisandra             234
                       ... 
Chowdeshwari Layout       1
6th Block Jayanagar       1
Ashraya Layout            1
Mudalpalaya               1
B Y Raveshwara Nagar      1
Name: location, Length: 1305, dtype: int64

Here we see that, **Whitefield and Sarjapur** have the max values i.e. they are most used or we can say that most of the Houses are in these areas.
Thus we can fill that single empty place in `location` column with any of the location i.e. Whitefield or Sarjapur

In [14]:
data['location'] = data['location'].fillna('Whitefield')

In [15]:
data['size'].value_counts()

2 BHK         5199
3 BHK         4310
4 Bedroom      826
4 BHK          591
3 Bedroom      547
1 BHK          538
2 Bedroom      329
5 Bedroom      297
6 Bedroom      191
1 Bedroom      105
8 Bedroom       84
7 Bedroom       83
5 BHK           59
9 Bedroom       46
6 BHK           30
7 BHK           17
1 RK            13
10 Bedroom      12
9 BHK            8
8 BHK            5
11 BHK           2
11 Bedroom       2
10 BHK           2
19 BHK           1
13 BHK           1
12 Bedroom       1
43 Bedroom       1
14 BHK           1
27 BHK           1
18 Bedroom       1
16 BHK           1
Name: size, dtype: int64

Here we see that, **2 BHK** is the most common or most occuring house size, thus we fill the missing ones with this.

In [16]:
data['size'] = data['size'].fillna('2 BHK')

In [17]:
data['size'].value_counts()

2 BHK         5215
3 BHK         4310
4 Bedroom      826
4 BHK          591
3 Bedroom      547
1 BHK          538
2 Bedroom      329
5 Bedroom      297
6 Bedroom      191
1 Bedroom      105
8 Bedroom       84
7 Bedroom       83
5 BHK           59
9 Bedroom       46
6 BHK           30
7 BHK           17
1 RK            13
10 Bedroom      12
9 BHK            8
8 BHK            5
11 BHK           2
11 Bedroom       2
10 BHK           2
19 BHK           1
13 BHK           1
12 Bedroom       1
43 Bedroom       1
14 BHK           1
27 BHK           1
18 Bedroom       1
16 BHK           1
Name: size, dtype: int64

In [18]:
data['bath'] = data['bath'].fillna(data['bath'].median())

In [19]:
data['bath'].value_counts()

2.0     6981
3.0     3286
4.0     1226
1.0      788
5.0      524
6.0      273
7.0      102
8.0       64
9.0       43
10.0      13
12.0       7
11.0       3
13.0       3
16.0       2
14.0       1
40.0       1
18.0       1
27.0       1
15.0       1
Name: bath, dtype: int64

In [20]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13320 non-null  object 
 1   size        13320 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13320 non-null  float64
 4   price       13320 non-null  float64
dtypes: float64(2), object(3)
memory usage: 520.4+ KB


Here we can see that all the missing values are filled. So now we can move further with our analysis.

In [21]:
data['bhk'] = data['size'].str.split().str.get(0).astype(int)
data['bhk']

0        2
1        4
2        3
3        3
4        2
        ..
13315    5
13316    4
13317    2
13318    4
13319    1
Name: bhk, Length: 13320, dtype: int32

In [22]:
data[data.bhk > 20]

location        size total_sqft  bath  price  bhk
1718  2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
4684                Munnekollal  43 Bedroom       2400  40.0  660.0   43

Since we can see that, there are only 2 records that have, `bhk` more then 20, which is of no use as there may not be any house with 43 rooms in it.

In [23]:
data['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

there are 9 unique records

In [24]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13320 non-null  object 
 1   size        13320 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13320 non-null  float64
 4   price       13320 non-null  float64
 5   bhk         13320 non-null  int32  
dtypes: float64(2), int32(1), object(3)
memory usage: 572.5+ KB


In [25]:
data['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

In [26]:
def convertRange(x):
    temp = x.split('-')
    if len(temp) == 2:
        return (float(temp[0]) + float(temp[1]))/2
    try:
        return float(x)
    except:
        return None

In [27]:
data['total_sqft'] = data['total_sqft'].apply(convertRange)

In [28]:
data.head()

location       size  total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2

## Price per Square feet

In [29]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13320 non-null  object 
 1   size        13320 non-null  object 
 2   total_sqft  13274 non-null  float64
 3   bath        13320 non-null  float64
 4   price       13320 non-null  float64
 5   bhk         13320 non-null  int32  
dtypes: float64(3), int32(1), object(2)
memory usage: 572.5+ KB


Now we can see that the, `total_sqft` column is in float64.

In [30]:
data['price_perSQ_ft'] = data['price']*100000 / data['total_sqft']

Doing this we get the price per sqr feet for each house.

In [31]:
data['price_perSQ_ft'].head()

0    3699.810606
1    4615.384615
2    4305.555556
3    6245.890861
4    4250.000000
Name: price_perSQ_ft, dtype: float64

In [32]:
data.describe()

total_sqft          bath         price           bhk  price_perSQ_ft
count  13274.000000  13320.000000  13320.000000  13320.000000    1.327400e+04
mean    1559.626694      2.688814    112.565627      2.802778    7.907501e+03
std     1238.405258      1.338754    148.971674      1.294496    1.064296e+05
min        1.000000      1.000000      8.000000      1.000000    2.678298e+02
25%     1100.000000      2.000000     50.000000      2.000000    4.266865e+03
50%     1276.000000      2.000000     72.000000      3.000000    5.434306e+03
75%     1680.000000      3.000000    120.000000      3.000000    7.311746e+03
max    52272.000000     40.000000   3600.000000     43.000000    1.200000e+07

In [33]:
data['location'].value_counts()

Whitefield              541
Sarjapur  Road          399
Electronic City         302
Kanakpura Road          273
Thanisandra             234
                       ... 
Chowdeshwari Layout       1
6th Block Jayanagar       1
Ashraya Layout            1
Mudalpalaya               1
B Y Raveshwara Nagar      1
Name: location, Length: 1305, dtype: int64

Since there are about 1305 records. and we cannot pass these many records to the model, thus we reduce them by putting "Others" in the columns that have value less than 10. For example : Agara Village has value 1, then we will replace this 1 and put other there.

In [34]:
data['location'] = data['location'].apply(lambda x : x.strip())
location_count = data['location'].value_counts()

In [35]:
location_count

Whitefield                              542
Sarjapur  Road                          399
Electronic City                         304
Kanakpura Road                          273
Thanisandra                             237
                                       ... 
RMV extension stage 2, rmv extension      1
Hoskote near                              1
Okalipura                                 1
Byrasandra Extension                      1
manyata tech park                         1
Name: location, Length: 1294, dtype: int64

In [36]:
location_count_less10 = location_count[location_count <= 10]
location_count_less10

Sector 1 HSR Layout                     10
Ganga Nagar                             10
Gunjur Palya                            10
Nagappa Reddy Layout                    10
Nagadevanahalli                         10
                                        ..
RMV extension stage 2, rmv extension     1
Hoskote near                             1
Okalipura                                1
Byrasandra Extension                     1
manyata tech park                        1
Name: location, Length: 1053, dtype: int64

there are about 1053 locations that have occured equal or less then 10 times in the data.

In [37]:
data['location'] = data['location'].apply(lambda  x : 'other' if x in location_count_less10 else x)


In [38]:
data['location'].value_counts()

other              2885
Whitefield          542
Sarjapur  Road      399
Electronic City     304
Kanakpura Road      273
                   ... 
Nehru Nagar          11
HAL 2nd Stage        11
Kodigehalli          11
Marsur               11
Banjara Layout       11
Name: location, Length: 242, dtype: int64

here we can see that, the size is reduced to 242 values.

In [39]:
data.describe()

total_sqft          bath         price           bhk  price_perSQ_ft
count  13274.000000  13320.000000  13320.000000  13320.000000    1.327400e+04
mean    1559.626694      2.688814    112.565627      2.802778    7.907501e+03
std     1238.405258      1.338754    148.971674      1.294496    1.064296e+05
min        1.000000      1.000000      8.000000      1.000000    2.678298e+02
25%     1100.000000      2.000000     50.000000      2.000000    4.266865e+03
50%     1276.000000      2.000000     72.000000      3.000000    5.434306e+03
75%     1680.000000      3.000000    120.000000      3.000000    7.311746e+03
max    52272.000000     40.000000   3600.000000     43.000000    1.200000e+07

If we look at this, we can see that the`total_sqft` have the min as `1.000000`, which is not possible. AS a house cannot be of 1sqft.

In [40]:
# here we fin that how much BHK shoul be there on 1 sqft
(data['total_sqft']/data['bhk']).describe()

count    13274.000000
mean       575.074878
std        388.205175
min          0.250000
25%        473.333333
50%        552.500000
75%        625.000000
max      26136.000000
dtype: float64

Here we can see that the `min`  is of 0.25 which is not possbile. Thus we will eliminate the flats/house which are having sqft less than 300.

In [41]:
data = data[((data['total_sqft']/data['bhk']) >= 300)]
data.describe()

total_sqft          bath         price           bhk  price_perSQ_ft
count  12530.000000  12530.000000  12530.000000  12530.000000    12530.000000
mean    1594.564544      2.559537    111.382401      2.650838     6303.979357
std     1261.271296      1.077938    152.077329      0.976678     4162.237981
min      300.000000      1.000000      8.440000      1.000000      267.829813
25%     1116.000000      2.000000     49.000000      2.000000     4210.526316
50%     1300.000000      2.000000     70.000000      3.000000     5294.117647
75%     1700.000000      3.000000    115.000000      3.000000     6916.666667
max    52272.000000     16.000000   3600.000000     16.000000   176470.588235

Here we cann see that the `min` total_sqft is 300.

In [42]:
data.shape

(12530, 7)

In [43]:
data.price_perSQ_ft.describe()

count     12530.000000
mean       6303.979357
std        4162.237981
min         267.829813
25%        4210.526316
50%        5294.117647
75%        6916.666667
max      176470.588235
Name: price_perSQ_ft, dtype: float64

the `max` value over here is not possible, so we have to remove this. Since the price_per_sqFt cannot be this high.

In [44]:
def removeSQFT_errors(df):
    df_op = pd.DataFrame()
    for key, subdf in df.groupby('location'):
        m = np.mean(subdf.price_perSQ_ft)
        st = np.std(subdf.price_perSQ_ft)
        gen_df = subdf[(subdf.price_perSQ_ft > (m-st)) & (subdf.price_perSQ_ft <= (m+st))]
        df_op = pd.concat([df_op, gen_df], ignore_index=True)
    return df_op
data = removeSQFT_errors(data)
data.describe()


total_sqft          bath         price           bhk  price_perSQ_ft
count  10301.000000  10301.000000  10301.000000  10301.000000    10301.000000
mean    1508.440608      2.471702     91.286372      2.574896     5659.062876
std      880.694214      0.979449     86.342786      0.897649     2265.774749
min      300.000000      1.000000     10.000000      1.000000     1250.000000
25%     1110.000000      2.000000     49.000000      2.000000     4244.897959
50%     1286.000000      2.000000     67.000000      2.000000     5175.600739
75%     1650.000000      3.000000    100.000000      3.000000     6428.571429
max    30400.000000     16.000000   2200.000000     16.000000    24509.803922

In [45]:
def removeBHK_errors(df):
    excluded = np.array([])
    for location, loc_df in df.groupby('location'):
        bhk_stats = {}
        for bhk, bhk_df in loc_df.groupby('bhk'):
            bhk_stats[bhk] = {
                'mean' : np.mean(bhk_df.price_perSQ_ft),
                'std' : np.std(bhk_df.price_perSQ_ft),
                'count' : bhk_df.shape[0]
            }
        for bhk, bhk_df in loc_df.groupby('bhk'):
            stats = bhk_stats.get(bhk-1)
            if stats and stats['count'] > 5:
                excluded = np.append(excluded, bhk_df[bhk_df.price_perSQ_ft < (stats['mean'])].index.values)
    return df.drop(excluded, axis='index')



In [46]:

data = removeBHK_errors(data)

In [47]:
data.shape

(7361, 7)

In [48]:
data

location       size  total_sqft  bath  price  bhk  \
0      1st Block Jayanagar      4 BHK      2850.0   4.0  428.0    4   
1      1st Block Jayanagar      3 BHK      1630.0   3.0  194.0    3   
2      1st Block Jayanagar      3 BHK      1875.0   2.0  235.0    3   
3      1st Block Jayanagar      3 BHK      1200.0   2.0  130.0    3   
4      1st Block Jayanagar      2 BHK      1235.0   2.0  148.0    2   
...                    ...        ...         ...   ...    ...  ...   
10292                other      2 BHK      1200.0   2.0   70.0    2   
10293                other      1 BHK      1800.0   1.0  200.0    1   
10296                other      2 BHK      1353.0   2.0  110.0    2   
10297                other  1 Bedroom       812.0   1.0   26.0    1   
10300                other      4 BHK      3600.0   5.0  400.0    4   

       price_perSQ_ft  
0        15017.543860  
1        11901.840491  
2        12533.333333  
3        10833.333333  
4        11983.805668  
...               ...  
10292     5833.333333  
10293    11111.111111  
10296     8130.081301  
10297     3201.970443  
10300    11111.111111  

[7361 rows x 7 columns]

In [49]:
data.drop(columns = ['size', 'price_perSQ_ft'], inplace=True)

## Cleaned Data

In [50]:
data.head()

location  total_sqft  bath  price  bhk
0  1st Block Jayanagar      2850.0   4.0  428.0    4
1  1st Block Jayanagar      1630.0   3.0  194.0    3
2  1st Block Jayanagar      1875.0   2.0  235.0    3
3  1st Block Jayanagar      1200.0   2.0  130.0    3
4  1st Block Jayanagar      1235.0   2.0  148.0    2

In [51]:
data.to_csv("cleaned_data.csv")

In [52]:
X = data.drop(columns = ['price'])
y = data['price']

In [53]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import  LinearRegression, Lasso, Ridge
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.metrics import  r2_score

In [54]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=0)

In [55]:
print(X_train.shape)
print(X_test.shape)

(5888, 4)
(1473, 4)


## Applying Linear Regression

In [56]:
column_trans = make_column_transformer((OneHotEncoder(sparse=False), ['location']),remainder='passthrough' )

In [57]:
scaler = StandardScaler()

In [58]:
lr = LinearRegression(normalize=True)

In [59]:
pipe = make_pipeline(column_trans, scaler, lr)

In [60]:
pipe.fit(X_train, y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()),
                ('linearregression', LinearRegression(normalize=True))])

In [61]:
y_pred_lr = pipe.predict(X_test)

In [62]:
r2_score(y_test, y_pred_lr)

0.8253671240835633

## Applying Lasso

In [63]:
lasso = Lasso()

In [64]:
pipe = make_pipeline(column_trans, scaler, lasso)

In [65]:
pipe.fit(X_train, y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()), ('lasso', Lasso())])

In [66]:
y_pred_lasso = pipe.predict(X_test)

In [67]:
r2_score(y_test, y_pred_lasso)

0.8146894751690389

## Applying Ridge

In [68]:
ridge = Ridge()

In [69]:
pipe = make_pipeline(column_trans, scaler, ridge)

In [70]:
pipe.fit(X_train, y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()), ('ridge', Ridge())])

In [71]:
y_pred_ridge = pipe.predict(X_test)

In [72]:
r2_score(y_test, y_pred_ridge)

0.8252348502290117

In [73]:
print("No Regularization : ", r2_score(y_test, y_pred_lr))
print("Lasso : ", r2_score(y_test, y_pred_lasso))
print("Ridge : ", r2_score(y_test, y_pred_ridge))

No Regularization :  0.8253671240835633
Lasso :  0.8146894751690389
Ridge :  0.8252348502290117


In [74]:
import pickle

In [75]:
pickle.dump(pipe, open('RidgeModel.pkl', 'wb'))